File: web/src/server/llm/retry.ts
export interface RetryOptions {
  maxRetries?: number
  baseDelayMs?: number
  maxDelayMs?: number
  shouldRetry?: (error: unknown) => boolean
}

export interface RetryResult<T> {
  value: T
  retryCount: number
}

export class RetryExhaustedError extends Error {
  readonly cause: unknown
  readonly retryCount: number

  constructor(error: unknown, retryCount: number) {
    super(error instanceof Error ? error.message : String(error))
    this.name = "RetryExhaustedError"
    this.cause = error
    this.retryCount = retryCount
  }
}

const DEFAULT_MAX_RETRIES = 2
const DEFAULT_BASE_DELAY_MS = 250
const DEFAULT_MAX_DELAY_MS = 2_000

export async function withRetry<T>(
  operation: () => Promise<T>,
  options: RetryOptions = {}
): Promise<RetryResult<T>> {
  const maxRetries = options.maxRetries ?? DEFAULT_MAX_RETRIES
  const baseDelayMs = options.baseDelayMs ?? DEFAULT_BASE_DELAY_MS
  const maxDelayMs = options.maxDelayMs ?? DEFAULT_MAX_DELAY_MS
  const shouldRetry = options.shouldRetry ?? isTransientNetworkError
  let retryCount = 0

  while (true) {
    try {
      return {
        value: await operation(),
        retryCount,
      }
    } catch (error) {
      if (retryCount >= maxRetries || !shouldRetry(error)) {
        throw new RetryExhaustedError(error, retryCount)
      }

      const delayMs = Math.min(baseDelayMs * 2 ** retryCount, maxDelayMs)
      retryCount += 1
      await delay(delayMs)
    }
  }
}

export function isTransientNetworkError(error: unknown): boolean {
  if (error instanceof TypeError) return true

  if (!error || typeof error !== "object") return false
  const maybeError = error as {
    code?: string
    retryable?: boolean
    transient?: boolean
    status?: number
  }

  if (maybeError.retryable || maybeError.transient) return true
  if (maybeError.status === 408 || maybeError.status === 429) return true
  if (typeof maybeError.status === "number" && maybeError.status >= 500) return true

  return [
    "ECONNRESET",
    "ECONNREFUSED",
    "EHOSTUNREACH",
    "ENETUNREACH",
    "ETIMEDOUT",
    "EAI_AGAIN",
  ].includes(maybeError.code ?? "")
}

function delay(ms: number): Promise<void> {
  if (ms <= 0) return Promise.resolve()
  return new Promise((resolve) => setTimeout(resolve, ms))
}

File: api/app/llm/retry.py
"""Retry with exponential backoff -- async port of web/src/server/llm/retry.ts."""
from __future__ import annotations

import asyncio
from dataclasses import dataclass
from typing import Any, Awaitable, Callable, Generic, TypeVar

T = TypeVar("T")

DEFAULT_MAX_RETRIES = 2
DEFAULT_BASE_DELAY_MS = 250
DEFAULT_MAX_DELAY_MS = 2_000

_TRANSIENT_UNIX_CODES = frozenset({
    "ECONNRESET",
    "ECONNREFUSED",
    "EHOSTUNREACH",
    "ENETUNREACH",
    "ETIMEDOUT",
    "EAI_AGAIN",
})


@dataclass(frozen=True)
class RetryOptions:
    max_retries: int = DEFAULT_MAX_RETRIES
    base_delay_ms: int = DEFAULT_BASE_DELAY_MS
    max_delay_ms: int = DEFAULT_MAX_DELAY_MS
    should_retry: Callable[[Exception], bool] | None = None


@dataclass(frozen=True)
class RetryResult(Generic[T]):
    value: T
    retry_count: int


class RetryExhaustedError(Exception):
    """Wraps the final cause after retries are exhausted."""

    def __init__(self, cause: Exception, retry_count: int) -> None:
        super().__init__(str(cause))
        self.cause = cause
        self.retry_count = retry_count


async def with_retry(
    operation: Callable[[], Awaitable[T]],
    options: RetryOptions | None = None,
) -> RetryResult[T]:
    opts = options or RetryOptions()
    should_retry = opts.should_retry or is_transient_network_error
    retry_count = 0

    while True:
        try:
            value = await operation()
            return RetryResult(value=value, retry_count=retry_count)
        except (asyncio.CancelledError, KeyboardInterrupt, SystemExit):
            raise
        except Exception as error:
            if retry_count >= opts.max_retries or not should_retry(error):
                raise RetryExhaustedError(error, retry_count) from error

            delay_ms = min(opts.base_delay_ms * (2 ** retry_count), opts.max_delay_ms)
            retry_count += 1
            if delay_ms > 0:
                await asyncio.sleep(delay_ms / 1000)


def is_transient_network_error(error: Any) -> bool:
    if isinstance(error, asyncio.TimeoutError):
        return True
    if getattr(error, "transient", False):
        return True
    if getattr(error, "retryable", False):
        return True

    status = getattr(error, "status", None)
    if status in (408, 429):
        return True
    if isinstance(status, int) and status >= 500:
        return True

    code = getattr(error, "code", None)
    if isinstance(code, str) and code in _TRANSIENT_UNIX_CODES:
        return True
    return False
